design-proposals/klip-59-json-functions.md

```diff
@@ -0,0 +1,244 @@
+* Calculate the length of a JSON array encoded in a string
+* Extract keys and records from a given JSON-formatted string
+* Concatenate JSON-formatted strings
+* Convert a ksqlDB record to JSON-formatted string
+
+## What is not in scope
+
+* A JSON datatype
+* Binary JSON support
+
+## Public APIs
+
+Users can use ksqlDB’s varchar type on columns that they know are JSON formatted. A JSON-formatted
+string can also be nested within a map or struct.
+
+### is_json
+
+```
+is_json(json_string)
+```
+
+Given a string, returns true if it can be parsed as a valid JSON value, false otherwise.
+
+#### Examples
+
+```
+is_json("[1, 2, 3]") // returns true
+is_json("{}") // returns true
+is_json("1") // returns true
+is_json("\"abc\"") // returns true
+is_json("null") // returns true
+is_json("") // returns false
+is_json("abc") // returns false
+is_json(NULL) // returns false
+```
+
+
+### json_array_length
+
+```
+json_array_length(json_string)
+```
+
+Given a string, parses it as a JSON value and returns the length of the top-level array. Returns
+`NULL` if the string can't be interpreted as a JSON array, i.e., it is `NULL` or it does not contain
+valid JSON, or the JSON value is not an array.
+
+#### Examples
+
+```
+json_array_length("[1, 2, 3]") // returns 3
+json_array_length("[1, [1, [2]], 3]") // returns 3
+json_array_length("[]") // returns 0
+json_array_length("{}") // returns NULL
+json_array_length("123") // returns NULL
+json_array_length("abc") // returns NULL
+json_array_length(NULL) // returns NULL
+```
+
+### json_keys
+
+```
+json_keys(json_string)
+```
+
+Given a string, parses it as a JSON object and returns a ksqlDB array of strings representing the
+top-level keys. Returns `NULL` if the string can't be interpreted as a JSON object, i.e., it is
+`NULL` or it does not contain valid JSON, or the JSON value is not an object.
+
+#### Examples
+
+```
+json_keys("{\"a\": \"abc\", \"b\": { \"c\": \"a\" }, \"d\": 1}") // returns ["a", "b", "c"]
+json_keys("{}") // returns []
+json_keys("[]") // returns NULL
+json_keys("") // returns NULL
+json_keys("123") // returns NULL
+json_keys("abc") // returns NULL
+json_keys(NULL) // returns NULL
+```
+
+### json_records
+
+```
+json_records(json_string)
+```
+
+Given a string, parses it as a JSON object and returns a ksqlDB array of two-element arrays
+representing the top-level keys and values. Returns `NULL` if the string can't be interpreted as a
+JSON object, i.e. it is `NULL` or it does not contain valid JSON, or the JSON value is not an object.
+
+#### Examples
+
+```
+json_records("{\"a\": \"abc\", \"b\": { \"c\": \"a\" }, \"d\": 1}") // returns [["a", "\"abc\""], ["b", "{ \"c\": \"a\" }"], ["d", "1"]]
+json_records("{}") // returns []
+json_records("[]") // returns NULL
+json_records("") // returns NULL
+json_records("123") // returns NULL
+json_records("abc") // returns NULL
+json_records(NULL) // returns NULL
+```
+
+#### Rejected alternatives
```

Comment:
I might be better to put all rejected alternative as the end of the KLIP instead of spreading it out for each function.

Reply:
I prefer keeping it in-line with each function closer to the reader's context

Reply:
Agree with @suhas-satish. In my view it is easier to reason about individual functions this way.